cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+                missing, resolved_bears, seen)
+            resolved_bears.append(bear)
+            seen.remove(bear)  # Already resolved, no candidate for circular dep
+
+        return resolved_bears
+
+    def resolve(self, bear_instance):
+        """
+        When a bear completes this method is called with the instance of that
+        bear. The method deletes this bear from the list of dependencies of
+        each bear in the dependency dictionary. It returns the bears which
+        have all of it's dependencies resolved.
+
+        :param bear_instance: The instance of the bear.
+        :return:              Returns a list of bears whose dependencies were
+                              all resolved and are ready to be scheduled.
+        """
+        return_bears = []
+        if bear_instance.type() in self.dependency_set:
+            for bear in self.dependency_dict:
+                if bear_instance.type() in self.dependency_dict[bear]:
+                    self.dependency_dict[bear].remove(bear_instance.__class__)
+                    if not self.dependency_dict[bear]:
+                        return_bears.append(bear)
+        for bear in return_bears:
+            del self.dependency_dict[bear]
+        return return_bears
+```
+
+# A Prototype
+
+A rough look at how the new local bear class implementation will look:
+
+```python
+class FileWiseBear:
+    def __init__(self, section, result_queue, file_set):
+        """
+        Constructs a new bear.
+
+        :param section:      The section object where bear settings are
+                             contained.
+        :param result_queue: The queue object for messages. Can be ``None``.
+        :param file_set:     A set of FileProxy objects.
+        """
+        self.section = section
+        self.result_queue = result_queue
+        self.file_set = file_set
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def execute_task(self, task, **kwargs):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        return list(self.analyze(task, **self.kwargs))
+
+    def analyze(self,
+                file_proxy,
+                *args,
+                dependency_results=None,
+                **kwargs):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        raise NotImplementedError("This function has to be implemented for a "
+                                  "runnable bear.")
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the files.
+        """
+        return self.file_set
+```
+
+A test bear:
+
+```python
+class TestBear(FileWiseBear):
+    def analyze(self, file_proxy):
```

Comment:
I would call it just file :D
because then you have stuff like `file.name`, `file.content`, `str(file)` which kind of makes it more readable imo :)